### DISCUSSION TITLE
How can I select a column on ColumnChart?

### DISCUSSION THREAD

**Discussion by fillipeoliveira-hotmart** [patrykandpatrick/vico#769] [Questions]
### Question

Hi,
Is possible to "select" a column on column chat like this?
<img width="431" alt="image" src="https://github.com/patrykandpatrick/vico/assets/51706700/ddd0f16e-f66d-4037-8d27-869ad1c73842">
In my usecase, I need to click on a bar and change the color of the selected. If I click on other, then i scroll to that position like 
<img width="430" alt="image" src="https://github.com/patrykandpatrick/vico/assets/51706700/32323de5-ec0c-45db-b8f4-ceacdf45c406">
In some cases I need to show a marker fixed on the selected position too.

The scroll on click I kind of manage to work adding a `markerVisibilityListener` and on `onShow` function scroll to position
```kotlin
scope.launch {
    graphScroll.animateScroll(Scroll.Absolute.x(targets.first().x, 0.5f))
}
```

That are my tests so far
```kotlin
@Composable
fun VicoGraph(modifier: Modifier = Modifier) {
    val scope = rememberCoroutineScope()
    val modelProducer = remember { CartesianChartModelProducer.build() }
    val graphScroll = rememberVicoScrollState()
    LaunchedEffect(Unit) {
        modelProducer.tryRunTransaction {
            columnSeries {
                series(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12)
            }
        }
    }
    CartesianChartHost(
        rememberCartesianChart(
            rememberColumnCartesianLayer(
                columnProvider = ColumnCartesianLayer.ColumnProvider.series(
                    rememberLineComponent(
                        color = Color(0xFF355CC0),
                        thickness = 32.dp,
                        shape = Shape.rounded(25)
                    )
                ),
                spacing = 24.dp,
            ),
            bottomAxis = rememberBottomAxis(
                valueFormatter = CartesianValueFormatter { value, _, _ ->
                    val months = DateFormatSymbols.getInstance(Locale("pt")).shortMonths
                    months[value.toInt() % 12]
                        .replace(".", "")
                        .replaceFirstChar {
                            if (it.isLowerCase()) {
                                it.titlecase(Locale("pt"))
                            } else {
                                it.toString()
                            }
                        }
                },
                label = rememberAxisLabelComponent(
                    color = Color(0xFF707780),
                    textSize = 12.sp,
                    padding = Dimensions.Empty,
                    margins = Dimensions.of(vertical = 16.dp)
                ),
                axis = null,
                tick = null,
                guideline = null,
            ),
        ),
        modelProducer,
        scrollState = graphScroll,
        marker = rememberMarker( // sample project marker
            labelPosition = DefaultCartesianMarker.LabelPosition.AbovePoint,
            showIndicator = false
        ),
        markerVisibilityListener = object : CartesianMarkerVisibilityListener {
            override fun onHidden(marker: CartesianMarker) {

            }

            override fun onShown(marker: CartesianMarker, targets: List<CartesianMarker.Target>) {
                val first = targets.first()
                scope.launch {
                    graphScroll.animateScroll(Scroll.Absolute.x(first.x, 0.5f))
                }
            }
        },
        modifier = modifier
    )
}
```

### Vico version(s)

2.0.0-alpha.21

### UI framework(s)

Jetpack Compose

**Comment by Gowsky** [comment]
Apologies for marking this as a duplicate without an update.

Please take a look at [`ColumnProvider`](https://patrykandpatrick.com/vico/api/2.0.0-alpha.21/core/com.patrykandpatrick.vico.core.cartesian.layer/-column-cartesian-layer/-column-provider/?query=interface%20ColumnProvider), you can use it to conditionally style columns using the data provided by `CartesianMarkerVisibilityListener`.

**Reply by fillipeoliveira-hotmart** [reply]
Thanks, I manage to work with the column provider.
I save `target.first().x` from the `CartesianMarkerVisibilityListener.onShown` and comparte with the `entry.x` on `ColumnProvider.getColumn`
I also save the marker and add it `persistentMarkers` to it work as a selected. The problem is that marker keep popping up when scroll the graph. Is possible to keep the scroll but not show the marker when try to scroll?

Here a sample

https://github.com/patrykandpatrick/vico/assets/51706700/d63e2631-453a-40c2-a4bb-a8c8fbe1763d

**Reply by cristianboicu01** [reply]
Did you manage to make it work? If so, can you please show how your implementation of ColumnCartesianLayer.ColumnProvider looks like? 

In my case, `getColumn` always has `seriesIndex = 0`. And it updates the color of all bars only when I click on the first bar (though I want to update only the bar I clicked on).

I am doing something wrong but I can't understand what exactly.

P.S. Figured it out. I was mistakenly considering `seriesIndex` as the index of the bar I click on. The actual index is `entry.x`

**Reply by cristianboicu01** [reply]
@Gowsky Do you have an answer or an idea on how to fix "the problem that marker keeps popping up when I scroll the graph"?

**Reply by weslley-telus** [reply]
@cristianboicu01 I've used the persistent markers to show with a state, I hope that can help you. 

```kotlin
 val scrubber = rememberMarker(
      color = settings.color,
      columnMargin = settings.columnMargin,
      entries = entries,
      scrubberFormatter = settings.scrubberFormatter
  )
  var rememberScrubberPosition by remember { mutableFloatStateOf(-1f) }
    
Chart(
        modifier = modifier,
        chart = columnChart(
            ... others properties
            axisValuesOverrider = settings.maxYValueOverrider,
            persistentMarkers = mapOf(rememberScrubberPosition to scrubber)
        ),
)
```

**Reply by cristianboicu01** [reply]
@weslley-telus Sorry but this does not help. I did manage to make selection work (to highlight a column when I click on it). The problem here is that when I scroll the chart, it highlights the bar I started to drag from even though I don't want to select another bar, just want to scroll in which case the selected bar should not change.

**Reply by Gowsky** [reply]
@fillipeoliveira-hotmart, @cristianboicu01, currently, if a chart has a marker, it always shows up at the touch point. We plan on adding customization options for this, which will enable you to achieve the result you've described.

**Reply by aelnonym** [reply]
@Gowsky Any updates on that? I'm facing the same issue of having the need for a column to be selected and for the chart to be scrollable in a Pager without selecting columns. Thanks in advance.

**Reply by Gowsky** [reply]
Hello, @aelnonym. We will share an update once the functionality is released.

**Reply by Gowsky** [reply]
[Vico v2.3.0-alpha.2](https://github.com/patrykandpatrick/vico/releases/tag/v2.3.0-alpha.2) introduces `CartesianMarkerController`.

**Comment by weslley-telus** [comment]
@PatrykGoworowski is it possible to do it using version 1.15.0?
What should I look for? 
I'm trying to do the same thing. 
I would appreciate it if you shared some code snippet about that.

**Reply by Gowsky** [reply]
Hi @weslley-telus, There's no such API in Vico 1. This is technically possible to do, but it would require lots of overrides in the existing API.
